src/main/java/com/edmunds/tools/databricks/maven/UpsertClusterMojo.java

```diff
@@ -0,0 +1,128 @@
+/*
+ *  Copyright 2019 Edmunds.com, Inc.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+package com.edmunds.tools.databricks.maven;
+
+import com.edmunds.rest.databricks.DTO.LibraryDTO;
+import com.edmunds.rest.databricks.DatabricksRestException;
+import com.edmunds.rest.databricks.request.CreateClusterRequest;
+import com.edmunds.rest.databricks.service.ClusterService;
+import com.edmunds.rest.databricks.service.LibraryService;
+import com.edmunds.tools.databricks.maven.model.ClusterTemplateModel;
+import com.edmunds.tools.databricks.maven.util.ObjectMapperUtils;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.maven.plugin.MojoExecutionException;
+import org.apache.maven.plugins.annotations.Mojo;
+import org.apache.maven.plugins.annotations.Parameter;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.edmunds.tools.databricks.maven.util.ClusterUtils.convertClusterNamesToIds;
+
+/**
+ * Cluster mojo, to perform databricks cluster upsert (create or update through recreation).
+ */
+@Mojo(name = "upsert-cluster")
+public class UpsertClusterMojo extends BaseDatabricksMojo {
+
+    private static final ObjectMapper OBJECT_MAPPER = ObjectMapperUtils.getObjectMapper();
+    private static final Map<String, String> SPARK_ENV_VARS = new HashMap<String, String>() {{
+        put("PYSPARK_PYTHON", "/databricks/python3/bin/python3");
+    }};
+
+    /**
+     * The databricks cluster json file that contains all of the information for how to create databricks cluster.
+     */
+    @Parameter(defaultValue = "${project.build.resources[0].directory}/databricks-plugin/databricks-cluster-settings.json", property = "dbClusterFile")
+    protected File dbClusterFile;
+
+    /**
+     * If true, command will fail if cluster with specified name already exists.
```

Comment:
im confused how this will be useful if it is for upserting? Can you describe use case?

Reply:
This flag protects us from unintentional upsert, according to Jira task: "If the cluster is already running, it will either fail or restart the cluster depending on additional options. Default would make sense to fail."

Can be changes anyway if you think it should be changed.

Reply:
ah I see, this is my fault for not providing clear definition.

Other users of this library who use this as part of CD pipeline preferred the library mojo not to terminate clusters as part of CD, but instead deploy changes and then restart manually later.
The thought was a similar view should be applied to these cluster setting upserts IF its possible. I know that some changes require termination of a cluster.

So the two different deployment options would be to:
- deploy all changes and restart
- deploy changes, but don't restart 

If its impossible to change cluster settings without a restart, then it is what it is.

Thoughts?

Reply:
Personally, I think if it requires a restart, then it should just do it. I would find it surprising to try to upsert on an existing job to NOT have it apply the changes and restart. I think this forces users to know that a restart is part of updating, which, is consistent with the way the databricks UI works anyways (e.g. if you make a cluster change, it will need to be restarted). I think the only caveat to this, is, if it has to delete the cluster and create new - that would be problematic. If that's necessary, then we would need some switch to protect unintended deletion.

Reply:
Looking at the docs and the code a little more, it is clear that it does not actually delete. So, I think the approach of removing the flag and just restarting on exist is my recommendation.